crates/sui-framework/sources/publisher.move

```diff
@@ -0,0 +1,145 @@
+/// in the module initializer.
+module sui::publisher {
+    use sui::object::{Self, UID};
+    use sui::tx_context::{TxContext, sender};
+    use std::ascii::{Self, String};
+    use std::type_name;
+    use std::vector;
+    use sui::types;
+
+    /// Tried to claim ownership using a type that isn't a one-time witness.
+    const ENotOneTimeWitness: u64 = 0;
+
+    /// ASCII character code for the `:` (colon) symbol.
+    const ASCII_COLON: u8 = 58;
+
+    /// This type can only be created in the transaction that
+    /// generates a module, by consuming its one-time witness, so it
+    /// can be used to identify the address that published the package
+    /// a type originated from.
+    struct Publisher has key, store {
+        id: UID,
+        type: String
+    }
+
+    /// Claim a Publisher object.
+    /// Requires a One-Time-Witness to
+    public fun claim<OTW: drop>(otw: OTW, ctx: &mut TxContext): Publisher {
+        assert!(types::is_one_time_witness(&otw), ENotOneTimeWitness);
+
+        Publisher {
+            id: object::new(ctx),
+            type: type_name::into_string(type_name::get<OTW>()),
+        }
+    }
+
+    /// Claim a Publisher object and send it to transaction sender.
+    /// Since this function can only be called in the module initializer,
+    /// the sender is the publisher.
+    public fun claim_and_keep<OTW: drop>(otw: OTW, ctx: &mut TxContext) {
+        sui::transfer::transfer(claim(otw, ctx), sender(ctx))
+    }
+
+    /// Destroy a Publisher object effectively removing all privileges
+    /// associated with it.
+    public fun burn(publisher: Publisher) {
+        let Publisher { id, type: _ } = publisher;
+        object::delete(id);
+    }
+
+    /// Check whether type belongs to the same package as the publisher object.
+    public fun is_package<T>(publisher: &Publisher): bool {
+        let this = ascii::as_bytes(&publisher.type);
+        let their = ascii::as_bytes(type_name::borrow_string(&type_name::get<T>()));
+
+        let i = 0;
+
+        // 40 bytes => length of the HEX encoded string
+        while (i < 40) {
+            if (vector::borrow<u8>(this, i) != vector::borrow<u8>(their, i)) {
+                return false
+            };
+
+            i = i + 1;
+        };
+
+        true
+    }
+
+    /// Check whether a type belogs to the same module as the publisher object.
+    public fun is_module<T>(publisher: &Publisher): bool {
+        if (!is_package<T>(publisher)) {
+            return false
+        };
+
+        let this = ascii::as_bytes(&publisher.type);
+        let their = ascii::as_bytes(type_name::borrow_string(&type_name::get<T>()));
+
+        // 42 bytes => length of the HEX encoded string + :: (double colon)
+        let i = 42;
+        loop {
+            let left = vector::borrow<u8>(this, i);
+            let right = vector::borrow<u8>(their, i);
+
+            if (left == &ASCII_COLON && right == &ASCII_COLON) {
```

Comment:
I would be inclined to put this logic for module name parsing in typename, since this surely won't be the last usage.

Reply:
Agreed! Makes a lot of sense!

Reply:
On the other hand I wouldn't make it a blocker for now since it will require changes in the std in the move repo.

Reply:
Should we mirror `stdlib` modules under the `sui` package, so we have a home for things that belong in a place like stdlib before we upstream them?

Reply:
@amnn @sblackshear 
One question I don't have an answer to is how can we know the length of the address type in Move? This module currently relies on addresses being 20 bytes (40 ASCII symbols). Should we include logic for 32-byte addresses as well?

Reply:
It's agood question, I think we should consider introducing a native `std::address::length()` to expose the `AccountAddress::LENGTH` constant within the VM (and this can supersede our own `sui::address::length()`.

Reply:
Added `std::address::length` with a hardcoded const.